### PR TITLE
Base64

### PR DESCRIPTION
**Pull request by mmatera**
Adding support to Export to and Import from Base64 binary to text encoding (see https://reference.wolfram.com/language/ref/format/Base64.html).

### CODE REVIEW

File: mathics/autoload/formats/Base64/Import.m [hunk 1]
```diff
@@ -0,0 +1,30 @@
+Begin["System`Convert`B64Dump`"]
+
+
+Options[B64Import] = {
+    "CharacterEncoding" :> $CharacterEncoding
+};
+
+B64Import[filename_, OptionsPattern[]] := 
+    Module[{strm,data, grid},
+	strm = OpenRead[filename];
```
Comment (sn6uv) [inline mathics/autoload/formats/Base64/Import.m]: `OpenRead` might fail (already handled for Export).

File: mathics/builtin/importexport.py [hunk 2]
```diff
@@ -981,3 +984,61 @@ def apply(self, filename, evaluation):
...
+        'System`Convert`B64Dump`B64Encode[expr_]'
+        if isinstance(expr,String):
+            stringtocodify = "\"" + expr.get_string_value() + "\""
+        else:
+            stringtocodify = Expression('ToString',expr).evaluate(evaluation).get_string_value()
+        return String(base64.b64encode(bytearray(stringtocodify, 'utf8')).decode('utf8'))
+            
+
+
+
```
Comment (sn6uv) [inline mathics/builtin/importexport.py]: too many newlines here

File: mathics/builtin/importexport.py [hunk 3]
```diff
@@ -981,3 +984,61 @@ def apply(self, filename, evaluation):
...
+    
+    context = "System`Convert`B64Dump`"
+    name = "B64Decode"
+
+    def apply(self, expr, evaluation):
+        'System`Convert`B64Dump`B64Decode[expr_String]'
+        try:
+            clearstring = base64.b64decode(bytearray(expr.get_string_value(), 'utf8')).decode('utf8')
+        except Exception as e:
+            print(e.__repr__())
```
Comment (sn6uv) [inline mathics/builtin/importexport.py]: remove print statement.

File: mathics/builtin/importexport.py [hunk 4]
```diff
@@ -981,3 +984,61 @@ def apply(self, filename, evaluation):
...
+            return Symbol("$Failed")
+        return Expression('ToExpression', String(six.text_type(clearstring))).evaluate(evaluation)
```
Comment (sn6uv) [inline mathics/builtin/importexport.py]: What happens if this fails, e.g. trying to decode b64 of `f[1`? or non-ascii chars.

File: mathics/builtin/importexport.py [hunk 5]
```diff
@@ -981,3 +984,61 @@ def apply(self, filename, evaluation):
...
+
+
+class B64Decode(Builtin):
+    """
+    <dl>
+    <dt> 'System`Convert`B64Dump`B64Decode[$string$]'
+    <dd>Decode  $string$ in Base64 coding to an expression.
+    </dl>
+
+    """
```
Comment (sn6uv) [inline mathics/builtin/importexport.py]: missing tests

File: mathics/builtin/importexport.py [hunk 6]
```diff
@@ -981,3 +984,61 @@ def apply(self, filename, evaluation):
...
+
+    >> System`Convert`B64Dump`B64Encode["Hello world"]
+     = IkhlbGxvIHdvcmxkIg==
+    >> System`Convert`B64Dump`B64Decode[%]
+     = Hello world
+    >> System`Convert`B64Dump`B64Encode[Integrate[f[x],{x,0,2}]]
+     = SW50ZWdyYXRlW2ZbeF0sIHt4LCAwLCAyfV0=
+    >> System`Convert`B64Dump`B64Decode[%]
+     = Integrate[f[x], {x, 0, 2}]
+    """
```
Comment (sn6uv) [inline mathics/builtin/importexport.py]: test for non-ascii char?

File: mathics/builtin/natlang.py [hunk 7]
```diff
@@ -1297,13 +1297,18 @@ class LanguageIdentify(Builtin):
...
         import langid  # see https://github.com/saffsd/langid.py
         # an alternative: https://github.com/Mimino666/langdetect
         import pycountry
         code, _ = langid.classify(text.get_string_value())
-        language = pycountry.languages.get(iso639_1_code=code)
+        try:
+            language = pycountry.languages.get(alpha_2=code)
+        except KeyError:
+            evaluation.message('LanguageIdentify', 'langnotfound', String(code))
+            return Symbol("$Failed")
```
Comment (sn6uv) [inline mathics/builtin/natlang.py]: is this change related?

File: mathics/builtin/importexport.py (hunk 5 above)
Comment (mmatera) [inline mathics/builtin/importexport.py]: Test was introduced in the function encode. Shall I repeat them here?

File: mathics/builtin/importexport.py (hunk 4 above)
Comment (mmatera) [inline mathics/builtin/importexport.py]: do you mean if the expression you get after decoding  is not a  valid mathics expression?

File: mathics/builtin/natlang.py (hunk 7 above)
Comment (mmatera) [inline mathics/builtin/natlang.py]: Nop... It was introduced to avoid an error in the travis tests, due to a former bug...

File: mathics/builtin/importexport.py (hunk 3 above)
Comment (mmatera) [inline mathics/builtin/importexport.py]: OK.

File: mathics/builtin/importexport.py (hunk 2 above)
Comment (mmatera) [inline mathics/builtin/importexport.py]: OK.

File: mathics/autoload/formats/Base64/Import.m (hunk 1 above)
Comment (mmatera) [inline mathics/autoload/formats/Base64/Import.m]: OK.

File: mathics/builtin/importexport.py (hunk 6 above)
Comment (mmatera) [inline mathics/builtin/importexport.py]: Done.

File: mathics/builtin/importexport.py (hunk 4 above)
Comment (mmatera) [inline mathics/builtin/importexport.py]: In that case, Expression(...) shows the message
Syntax::sntxi: Incomplete expression; more input is needed (line 1 of "")
and returns 
$Failed

Now, I catch this output and in case of fail, an specific message is shown.

File: mathics/builtin/importexport.py (hunk 5 above)
Comment (mmatera) [inline mathics/builtin/importexport.py]: Now I have  included specific error handling tests there.

File: mathics/builtin/importexport.py (hunk 4 above)
Comment (mmatera) [inline mathics/builtin/importexport.py]: I realized that is better to codify strings and not expressions. This is the standard behaviour in mma, and is needed if you want to use this function to codify binary files.

**Comment by mmatera** [comment]
I realised I forgot to mark the comments as resolved...
